### PR TITLE
prov/rxm: add FI_PEER support to rxm

### PR DESCRIPTION
**Pull request by aingerson**
Add rxm support for FI_PEER (srx, cq, and cntr) and FI_AV_USER_ID to be used with the link provider

### CODE REVIEW

**Comment by aingerson** [comment]
(will move first 3 commits into separate PR, leaving here for testing)

File: prov/rxm/src/rxm_ep.c [hunk 1]
```diff
@@ -1525,9 +1199,108 @@ static int rxm_ep_enable_check(struct rxm_ep *rxm_ep)
...
+			   struct fid_ep **rx_ep, void *context)
+{
+	struct rxm_domain *rxm_domain;
+
+	rxm_domain = container_of(domain, struct rxm_domain,
+				  util_domain.domain_fid);
+
+	if (attr->op_flags & FI_PEER) {
+		rxm_domain->srx = ((struct fi_peer_srx_context *)
+					(context))->srx;
```
Comment (j-xiong) [inline prov/rxm/src/rxm_ep.c:1279]: Should we check if `srx` has already been set?

File: prov/rxm/src/rxm_domain.c [hunk 2]
```diff
@@ -221,6 +221,25 @@ static struct fi_ops_av_owner rxm_av_owner_ops = {
...
+	struct rxm_rx_buf *rx_buf = rx_entry->peer_context;
+
+	return rx_buf->conn->peer->fi_addr;
+}
+
+static void rxm_foreach_ep(struct util_av *av, struct util_ep *ep)
+{
+	struct rxm_ep *rxm_ep;
+	struct fid_peer_srx *peer_srx;
+
```
[the remark this answers is not in the document]
Comment (aingerson) [reply prov/rxm/src/rxm_domain.c:235]: I don't think it will complain about unused params. Do you want me to add it anyway?

**Comment by aingerson** [comment]
@j-xiong Almost all of the alignment you pointed out are places where I did it on purpose to maintain the 80 character limit. Do you want me to change all of those to adding a new line after the open bracket? It looks weird to me and I don't think we do it most places in libfabric but I can change them all if you like.

**Comment by j-xiong** [comment]
Yes, I prefer making the changes if possible.

________________________________

File: prov/rxm/src/rxm_ep.c [hunk 3]
```diff
@@ -1596,7 +1395,7 @@ static int rxm_ep_bind(struct fid *ep_fid, struct fid *bfid, uint64_t flags)
 	rxm_ep = container_of(ep_fid, struct rxm_ep, util_ep.ep_fid.fid);
 
         ret = ofi_ep_bind(&rxm_ep->util_ep, bfid, flags);
-	if (ret)
+	if (ret && ret != -FI_EINVAL)
```
[the remark this answers is not in the document]
Comment (aingerson) [reply prov/rxm/src/rxm_ep.c:1398]: Yes. But I agree it looks weird. I can code it differently to be more intuitive

**Comment by aingerson** [comment]
Update -
Fixed alignment and formatting issues
Rewrote bind path
Check for already existing srx on create and bind paths

**Review by j-xiong — APPROVED** [review]
Thanks.

**Comment by j-xiong** [comment]
@shijin-aws Is the AWS CI failure related?

**Comment by j-xiong** [comment]
bot:aws:retest

**Comment by j-xiong** [comment]
2 mpichtestsuite test cases failed with Intel CI over verbs_rxm:
```
not ok 417 - ./pt2pt/icprobe 4 # time=180.3842689991
APPLICATION TIMED OUT, TIMEOUT = 180s

not ok 418 - ./pt2pt/icprobe 4 # time=180.858914136887
APPLICATION TIMED OUT, TIMEOUT = 180s
```

**Pull request merged**